### PR TITLE
Allow rendering response data in the form success/error messages

### PR DESCRIPTION
**Pull request by mkhatib**
ITI: #3343 

- [x] Move XHR changes to a different PR

### CODE REVIEW

**Comment by cramforce** [comment]
@dvoytenko should do a quick pass for mustache integration.

File: extensions/amp-form/0.1/amp-form.js [hunk 1]
```diff
@@ -88,11 +92,14 @@ export class AmpForm {
...
-            this.setState_(FormState_.SUBMIT_ERROR);
-            rethrowAsync('Form submission failed:', error);
-          });
+      }).then(response => {
+        this.setState_(FormState_.SUBMIT_SUCCESS);
+        this.renderTemplate_(FormState_.SUBMIT_SUCCESS, response);
+      }).catch(error => {
+        this.setState_(FormState_.SUBMIT_ERROR);
+        this.renderTemplate_(FormState_.SUBMIT_ERROR, error).then(
+            rethrowAsync.bind(null, 'Form submission failed:', error));
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Why is this in a then block? Can't it be done sync?

File: extensions/amp-form/0.1/amp-form.js [hunk 2]
```diff
@@ -109,6 +116,26 @@ export class AmpForm {
...
+
+  /**
+   * @param {string} state
+   * @param {!Object} data
+   * @private
+   */
+  renderTemplate_(state, data) {
+    const container = this.form_.querySelector(`[${state}]`);
+    if (container) {
+      const previousRender = container.querySelector(`.${state}-message`);
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: How about removing this element during the submit? (It's also fix transitioning from error -> success and success -> error that this has).

From a UI perspective, if the POST takes a few seconds to finish, I could be staring at an error state thinking my input is bad.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Makes sense. Will fix this.

File: extensions/amp-form/0.1/amp-form.js (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: I found this work better for testing. I don't really understand exactly why but basically If I rethrowAsync sync after the `this.renderTemplate` call, the test always fail with the error being thrown, if I try to expect that test to eventually throw it would catch the error but it doesn't seem to test the last block of the test which does the actual expectations inside the `timer.promise(0).then`.

I can try and debug this some more if you think we should do this sync.

Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Likely because you are "catching" the `UnhandledRejection` when thrown async.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: How do we avoid this? I might need to re-write the way I am testing this. Will try to do some more tests here.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Fixed this by changing `timer.promise(20)` to `timer.promise(0)` in the test.

File: extensions/amp-form/0.1/amp-form.js (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Mind if I do this in a separate PR. The other PR I have open introduce the `this.state_` property and don't wanna introduce it again in this one.

Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Sure, that's fine with me.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Filed a bug #3587 to track this.

File: extensions/amp-form/0.1/amp-form.js (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done.

**Comment by mkhatib** [comment]
PTAL

File: examples/forms.amp.html [hunk 3]
```diff
@@ -8,12 +8,15 @@
...
     <script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js"></script>
+    <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.1.js"></script>
     <style amp-custom>
         input:invalid {
             background-color: #dc4e41;
         }
 
-        form .form-message {
+        form .form-message,
+        form [submit-success],
```
Comment (dvoytenko) [inline examples/forms.amp.html]: If we introduce `submit-error` and such attributes - we have to configure CSS for them in our CSS, right? Otherwise, it'd be illogical to whitelist these attributes in validator but not give them any default meaning.

File: extensions/amp-form/0.1/amp-form.js [hunk 4]
```diff
@@ -109,6 +116,39 @@ export class AmpForm {
...
+   * @private
+   */
+  renderTemplate_(state, data) {
+    const container = this.form_.querySelector(`[${state}]`);
+    if (container) {
+      // TODO(#3587): Move the cleanup to do during submission.
+      this.cleanupRenderedTemplate_(state);
+      return this.templates_.findAndRenderTemplate(container, data)
+          .then(rendered => {
+            rendered.classList.add(`${state}-message`);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Do we have to have this class? Is there any way to do it some other way? E.g. we can use `i-amp-rendered` attribute or such? This way we won't have any ambiguity of conflict against the class name?

File: extensions/amp-form/0.1/amp-form.js [hunk 5]
```diff
@@ -109,6 +116,39 @@ export class AmpForm {
...
+  /**
+   * @param {string} state
+   * @private
+   */
+  cleanupRenderedTemplate_(state) {
+    const container = this.form_.querySelector(`[${state}]`);
+    if (!container) {
+      return;
+    }
+    const previousRender = container.querySelector(`.${state}-message`);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: If we do `i-amp-rendered` or such, we can then use a more optimized `dom.childElementByAttr`

File: extensions/amp-form/0.1/amp-form.js [hunk 6]
```diff
@@ -88,11 +92,14 @@ export class AmpForm {
...
-          .catch(error => {
-            this.setState_(FormState_.SUBMIT_ERROR);
-            rethrowAsync('Form submission failed:', error);
-          });
+      }).then(response => {
+        this.setState_(FormState_.SUBMIT_SUCCESS);
+        this.renderTemplate_(FormState_.SUBMIT_SUCCESS, response);
+      }).catch(error => {
+        this.setState_(FormState_.SUBMIT_ERROR);
+        this.renderTemplate_(FormState_.SUBMIT_ERROR, error);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: How can `error` be actually data here? This is a bit of a XSS - since you are exposing an object with the state you do not fully control to a template. Let's find another way to pass data in case of an error.

File: examples/forms.amp.html (hunk 3 above)
Comment (mkhatib) [inline examples/forms.amp.html]: Yes I think we should. I think it makes sense in these two cases to manage display for the user as well. So basically moving these styles to be provided by amp by default. Des that make sense?

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Sure, `i-amp-rendered` should work similarly.

File: extensions/amp-form/0.1/amp-form.js (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Yeah, we don't currently pass real data and just use the error instance that xhr-impl throws. The current xhr setup ignores the body of errors statuses. 

Can we change the implementation for fetchJson? Instead of throwing an error in `xhr.assertSuccess` we instead return a rejected promise with the response?

Would this be a way to do this?
```
    return this.fetchAmpCors_(input, init).then(response => {
      if (!isSuccess(response)) {
        return Promise.reject({error: new Error('message'), response: response});
      }
      return response.json();
    })
```

And then the invoker would do something like the following:
```javascript
      fetchJson().then(...).catch(errorData => {
        this.setState_(FormState_.SUBMIT_ERROR);
        errorData.response.json().then(data =>
            this.renderTemplate_(FormState_.SUBMIT_ERROR, data));
        rethrowAsync('Form submission failed:', errorData.error);
      });
```

What do you think?

File: examples/forms.amp.html (hunk 3 above)
Comment (dvoytenko) [inline examples/forms.amp.html]: Yes, specifically `form [submit-success]` and `form [submit-error]`, right?

Comment (mkhatib) [inline examples/forms.amp.html]: Yes exactly.

File: extensions/amp-form/0.1/amp-form.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I think it'd not be very typical to expect the error response to return JSON. But we can certainly try to parse it and include it in the rejection of the promise. E.g. `error.responseJson` or such. Then we can use it here. But I don't think it'd be a real-life usecase. Most likely one would return a successful structure that would describe erroneous situation and use template to show it as such. E.g.

```
<div submit-success>
  <template>
   {#error}
     show error
  {/#error}
  otherwise show success message
```

This case we already support with this code. If you'd like to also support the JSON in the error response - I'm fine if we do that as well. Up to you. But I don't think it'd be super common.

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done. Let me know if this looks good.

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done.

File: examples/forms.amp.html (hunk 3 above)
Comment (mkhatib) [inline examples/forms.amp.html]: Should I move the styles into `amp-form.css` or these should be part of `amp.css`? Should I do it in the PR?

File: extensions/amp-form/0.1/amp-form.js (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Also, I can do this in a separate PR if you'd like - I need to update couple of xhr test cases that this breaks and check other existing usage just to make sure this works well.

File: examples/forms.amp.html (hunk 3 above)
Comment (dvoytenko) [inline examples/forms.amp.html]: At the very least, `form [..] {display: none}` should be in `amp.css` to avoid FOUC.

File: extensions/amp-form/0.1/amp-form.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Error path in separate PR is certainly fine.

File: extensions/amp-form/0.1/amp-form.js [hunk 7]
```diff
@@ -88,11 +92,14 @@ export class AmpForm {
...
-          .catch(error => {
-            this.setState_(FormState_.SUBMIT_ERROR);
-            rethrowAsync('Form submission failed:', error);
-          });
+      }).then(response => {
+        this.setState_(FormState_.SUBMIT_SUCCESS);
+        this.renderTemplate_(FormState_.SUBMIT_SUCCESS, response);
+      }).catch(error => {
+        this.setState_(FormState_.SUBMIT_ERROR);
+        this.renderTemplate_(FormState_.SUBMIT_ERROR, error.responseJson);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Should probably be `responseJson || {}`? Looks like `data` is always required there.

**Comment by dvoytenko** [comment]
So, this is LGTM. Just one request to fallback `error.responseJson || {}`. And please move XHR changes themselves into a separate PR.

File: extensions/amp-form/0.1/amp-form.js (hunk 7 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: I've used default parameter syntax in the `renderTemplate_(state, data = {})`, is that ok?

Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Hmm. I don't think that what it does. The default parameter would take on the specified value when this parameter is not specified at all, e.g. `renderTemplate_(FormState.X)`. In this case you are passing `undefined` as the argument value.

File: src/service/xhr-impl.js [hunk 8]
```diff
@@ -344,17 +344,26 @@ function isRetriable(status) {
...
-    const err = user.createError(`HTTP error ${response.status}`);
-    if (isRetriable(response.status)) {
-      err.retriable = true;
+  return new Promise((resolve, reject) => {
+    if (!(response.status >= 200 && response.status < 300)) {
+      const err = user.createError(`HTTP error ${response.status}`);
+      if (isRetriable(response.status)) {
+        err.retriable = true;
+      }
+      return response.json().then(json => {
```
Comment (dvoytenko) [inline src/service/xhr-impl.js]: (You'll probably move this to a separate PR, but...) Before attempting to do `json()` here, we need to check that the `Content-Type` is actually JSON. The error response doesn't have to follow the same protocol as the successful response and typically doesn't do so.

File: extensions/amp-form/0.1/amp-form.js (hunk 7 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: You're right, sorry misunderstood the default param. Will fix it.

File: src/service/xhr-impl.js (hunk 8 above)
Comment (mkhatib) [inline src/service/xhr-impl.js]: Sent a separate PR #3669

Comment (mkhatib) [inline src/service/xhr-impl.js]: Done.

File: extensions/amp-form/0.1/amp-form.js (hunk 7 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Fixed.

File: extensions/amp-form/0.1/amp-form.js (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done.

File: examples/forms.amp.html (hunk 3 above)
Comment (mkhatib) [inline examples/forms.amp.html]: Done.

Tracking adding `display:block` to these containers in `amp-form.css` in the ITI PR. Will add in next PRs.
